bin/run_clang_tidy_check

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/python3
+# Runs clang-tidy over the bob_robotics repo to find the specified warnings and
+# converts the results into a Jenkins-readable XML file. One or more
+# compile_commands.json files should be present somewhere in the tree (they
+# don't have to be committed to git). These can be generated by CMake by passing
+# the argument -DCMAKE_EXPORT_COMPILE_COMMANDS=1.
+#
+# If the argument --generate-fixes is passed to this script, then it will
+# deposit a series of YAML files into ./clang_tidy_fixes. These can then be
+# applied like so:
+#       clang-apply-replacements ./clang_tidy_fixes
+#
+# Note that only a small subset of the checks are actually capable of generating
+# fixes at present (though most of the modernize-* checks seem to work).
+
+import collections
+import itertools
+import json
+import logging
+import multiprocessing as mp
+import os
+import re
+import subprocess
+import sys
+from xml.sax.saxutils import escape
+
+INCLUDE_CHECKS = (
+    'bugprone-*',
+    'clang-analyzer-*',
+    'concurrency-*',
+    'misc-*',
+    'modernize-*',
+    'openmp-*',
+    'performance-*',
+)
+
+# Some of these are a bit overzealous...
+EXCLUDE_CHECKS = (
+    '*osx.*',
+    'bugprone-branch-clone',
+    'bugprone-exception-escape',
+    'bugprone-integer-division',
+    'bugprone-narrowing-conversions',
+    'clang-analyzer-webkit*',
+    'misc-non-private-member-variables-in-classes',
+    'modernize-avoid-c-arrays',
+    'modernize-concat-nested-namespaces',
+    'modernize-pass-by-value',
+    'modernize-use-nodiscard',
+    'modernize-use-auto',
+    'modernize-use-emplace',
+    'modernize-use-override',
+    'modernize-use-trailing-return-type',
+    'modernize-raw-string-literal',
+    'modernize-return-braced-init-list',
+    'openmp-use-default-none',
+    'performance-type-promotion-in-math-fn',
+)
+
+# Create a `ErrorDescription` tuple with all the information we want to keep.
+ErrorDescription = collections.namedtuple(
+    'ErrorDescription', 'file line column error error_identifier description')
+
+# This class was adapted from here: https://github.com/PSPDFKit-labs/clang-tidy-to-junit
+class ClangTidyConverter:
```

Comment:
This is cool but presenting the clang-tidy output via junit seems weird to me as they appear in the tests section of the jenkins report - wouldn't it make more sense to show these through warnings NG (which claims to support clang-tidy output)

Reply:
Hmm... I didn't notice that warnings NG supported it! I basically stole the idea and most of the code from here: https://pspdfkit.com/blog/2018/using-clang-tidy-and-integrating-it-in-jenkins/

I'm guessing that they were converting to junit files either a) because warnings NG wasn't a thing when they did it or b) the output format is more useful.

I'll maybe try pushing a random error to the tree to see how warnings NG formats it. But if it looks ok, then that's definitely more sensible.